Index: sys/arm/allwinner/if_awgreg.h
===================================================================
--- sys/arm/allwinner/if_awgreg.h	(revision 323952)
+++ sys/arm/allwinner/if_awgreg.h	(working copy)
@@ -115,9 +115,9 @@
 #define	EMAC_MII_DATA		0x4c
 #define	EMAC_ADDR_HIGH(n)	(0x50 + (n) * 8)
 #define	EMAC_ADDR_LOW(n)	(0x54 + (n) * 8)
-#define	EMAC_TX_DMA_STA		0x80
-#define	EMAC_TX_DMA_CUR_DESC	0x84
-#define	EMAC_TX_DMA_CUR_BUF	0x88
+#define	EMAC_TX_DMA_STA		0xb0
+#define	EMAC_TX_DMA_CUR_DESC	0xb4
+#define	EMAC_TX_DMA_CUR_BUF	0xb8
 #define	EMAC_RX_DMA_STA		0xc0
 #define	EMAC_RX_DMA_CUR_DESC	0xc4
 #define	EMAC_RX_DMA_CUR_BUF	0xc8
